src/compositor/windowpixmapitem.cpp

```diff
@@ -542,131 +601,59 @@ void WindowPixmapItem::handleWindowSizeChanged()
 void WindowPixmapItem::itemDestroyed(QObject *)
 {
     m_item = 0;
-    if (!m_haveSnapshot) {
-        m_hasPixmap = false;
-        emit hasPixmapChanged();
+    m_surface.reset();
+    m_captureSurface.reset();
+    disconnect(window(), &QQuickWindow::beforeRendering, this, &WindowPixmapItem::captureSnapshot);
+    m_clearCapture = false;
+    update();
+    updateHasPixmap();
+}
+
+bool WindowPixmapItem::event(QEvent *event)
+{
+    if (event->type() == QEvent::User) {
+        m_clearNotified = false;
+        if (m_clearCapture) {
+            m_clearCapture = false;
+            m_captureSurface.reset();
+        }
+        updateHasPixmap();
+        return true;
+    } else {
+        return QQuickItem::event(event);
     }
 }
 
 QSGNode *WindowPixmapItem::updatePaintNode(QSGNode *oldNode, UpdatePaintNodeData *)
 {
-    SurfaceNode *node = static_cast<SurfaceNode *>(oldNode);
-
-    if (m_item == 0 && !m_haveSnapshot) {
-        if (node)
-            node->setTextureProvider(0, false);
-        delete node;
-        return 0;
+    if (m_captureSurface && !m_clearCapture) {
+        captureSnapshot();
```

Comment:
ShaderEffectSource would have been perfect for this approach, and it so nearly worked.

But it would seem QWaylandBufferRef does nothing to stop the underlying wayland buffer from being destroyed when the client window is hidden.  An outstanding GL texture keeps the underlying gralloc buffer alive for a time but that's destroyed when the scene graph is invalidated, so it looks as though its worked but then display turns off and it turns back on again all the buffers are gone.